File: 0099_RecoverBinarySearchTree/RecoverBinarySearchTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def recoverTree(self, root: TreeNode) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        s = list()
        prev = None
        x = None
        y = None
        while s or root:
            while root:
                s.append(root)
                root = root.left
            
            root = s.pop(-1)
            if prev and root.val < prev.val: 
                y = root
                if not x:
                    x = prev
                else: 
                    break
            prev = root
            root = root.right
        
        x.val, y.val = y.val, x.val

File: 0099_RecoverBinarySearchTree/RecoverBinarySearchTree.cxx
#include<vector>
#include<stack>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    void recoverTree(TreeNode* root) {
        stack<TreeNode* > s;
        TreeNode* x = nullptr;
        TreeNode* y = nullptr;
        TreeNode* pred = nullptr;
        while(!s.empty() || root){
            while( root ){
                s.push(root);
                root = root->left;
            }
            root = s.top();
            s.pop();
            if( pred && root->val < pred->val){
                y = root;
                if(!x){
                    x = pred;
                }
                else break;
            }
            pred = root;
            root = root->right;
        }
        swap(x->val, y->val);

    }
};